src/Parser/Hunter/Marksmanship/Modules/Traits/Bullseye.js

```diff
@@ -0,0 +1,103 @@
+import React from 'react';
+import Analyzer from 'Parser/Core/Analyzer';
+import SPELLS from 'common/SPELLS';
+import StatisticBox from "Main/StatisticBox";
+import SpellIcon from "common/SpellIcon";
+import { formatNumber, formatPercentage } from "common/format";
+import CooldownTracker from '../Features/CooldownTracker';
+
+class Bullseye extends Analyzer{
+  executeTimestamp;
+  bullseyeResets = 0; //only resets when boss < 20% health, so resets we can confirm shouldn't have happened
+  bullseyeInstances = [];
+  saveForExecuteThreshhold;
+  bossIDs = [];
+
+  static dependencies = {
+    cooldownTracker : CooldownTracker,
+  }
+  on_initialized(){
+    this.owner.report.enemies.forEach(enemy => {
+      if (enemy.fights[0].id === this.owner.fight.id && enemy.type === "Boss"){ //ensure that we only count Bosses for checking if bullseye was reset during execute
+        this.bossIDs.push(enemy.id);
+      }
+    });
+  }
+  on_byPlayer_applybuff(event){
+    if (event.ability.guid !== SPELLS.BULLSEYE_TRAIT.id){
+      return;
+    }
+    this.bullseyeInstances.push({'start':event.timestamp - this.owner.fight.start_time});
+  }
+
+  on_byPlayer_applybuffstack(event) {
+    if (event.ability.guid !== SPELLS.BULLSEYE_TRAIT.id){
+      return;
+    }
+    if (event.stack === 30){
+      this.bullseyeInstances[this.bullseyeInstances.length-1].maxStacksTimestamp = event.timestamp - this.owner.fight.start_time;
+    }
+  }
+
+  on_byPlayer_removebuff(event){
+    if (event.ability.guid !== SPELLS.BULLSEYE_TRAIT.id){
+      return;
+    }
+    if (this.executeTimestamp){
+      this.bullseyeResets ++;
+      this.bullseyeInstances[this.bullseyeInstances.length-1].duringExecute = true;
+    }
+    this.bullseyeInstances[this.bullseyeInstances.length-1].end = event.timestamp - this.owner.fight.start_time;
+  }
+
+  on_byPlayer_damage(event){
+    if(event.targetInstance === undefined && event.maxHitPoints && this.bossIDs.indexOf(event.targetID) !== -1){
+      if ((event.hitPoints / event.maxHitPoints * 10000)/100 <= 20 && !this.executeTimestamp){
```

Comment:
I think it would be cleaner to not multiply by 10,000 and then divide by 100 and instead just check for 0.2 (20%).

Reply:
With regards to the 20%: Please extract magic numbers to constants at the top of the file with appropriate names.